pkg/datasource/datasource.go

```diff
@@ -65,35 +100,51 @@ type DataSource interface {
 	// AddField adds a field as a new payload
 	AddField(fieldName string, kind api.Kind, options ...FieldOption) (FieldAccessor, error)
 
-	// NewData builds a new data structure that can be written to
-	NewData() Data
+	// NewPacketSingle builds a new PacketSingle that can be written to
+	NewPacketSingle() PacketSingle
+	// NewPacketSingleFromRaw builds a new PacketSingle from a raw bytes slice coming from protobuf
+	NewPacketSingleFromRaw(b []byte) (PacketSingle, error)
 
-	// NewDataFromRaw builds a new data structure from a raw bytes slice coming from protobuf
-	NewDataFromRaw(b []byte) (Data, error)
+	// NewPacketArray and NewPacketArrayFromRaw are analogous to NewPacketSingle and NewPacketSingleFromRaw, but for
+	// PacketArray
+	NewPacketArray() PacketArray
+	NewPacketArrayFromRaw(b []byte) (PacketArray, error)
```

Comment:
It feels a bit weird to have two set of functions to allocate Data, also that you can call NewPacketSingle in an array data source and viceversa. I was wondering if it was possible to split this interface by having DataSourceSingle and DataSourceArray, but it quickly became a mess. I think this solution is fine, but I'd like to have some more checks around to avoid the user calling the wrong function in a non-valid type, for instance, to call SubscribeArray() on a TypeSingle datasource, that should fail.

Reply:
I added the error and some test to verify the proper behaviour for creating new packets and subscribing to data, dataArray and packet.